voxels/Voxelize.cpp

```diff
@@ -371,13 +279,514 @@ std::vector<std::byte> raw_voxelize_obj(std::string_view filepath,
+                    auto texture = loaded_textures_[face_mat_id];
+                    texture_pixels = std::get<0>(texture);
+                    texture_width = std::get<1>(texture);
+                    texture_height = std::get<2>(texture);
+                } else {
+                    const tinyobj::material_t &mat = materials[face_mat_id];
+                    const std::string &diffuse_texname = mat.diffuse_texname;
+                    if (diffuse_texname.empty()) {
+                        has_materials_ = false;
+                    } else {
+                        int channels;
+                        texture_pixels =
+                            stbi_load((directory + diffuse_texname).c_str(),
+                                      &texture_width, &texture_height,
+                                      &channels, STBI_rgb_alpha);
+                        if (!texture_pixels) {
+                            std::stringstream ss;
+                            ss << "Couldn't load texture from the following "
+                                  "path: ";
+                            ss << (directory + diffuse_texname);
+                            ss << "\n";
+                            ASSERT(false, ss.str());
+                        }
+                        loaded_textures_[face_mat_id] = {
+                            texture_pixels, texture_width, texture_height};
+                    }
+                }
+            }
+        }
+    }
+}
+
+Model::~Model() {
+    for (auto [_, tex] : loaded_textures_) {
+        stbi_image_free(std::get<0>(tex));
+    }
+}
+
+Voxelizer::Voxelizer(std::string_view filepath, float voxel_size) : 
+    model_{filepath}, current_memory_usage_{0} 
+{
+    glm::vec3 max = model_.get_max();
+    glm::vec3 min = model_.get_min();
+    width_ = static_cast<uint32_t>(ceil((max.x - min.x) / voxel_size));
+    height_ = static_cast<uint32_t>(ceil((max.y - min.y) / voxel_size));
+    depth_ = static_cast<uint32_t>(ceil((max.z - min.z) / voxel_size));
+    std::cout << "Chunk size: (" 
+              << width_ << ", " << height_ << ", " << depth_ 
+              << ")\n";
+
+    // Allocate the number of voxel chunks that we'll need.
+    chunks_width_ = static_cast<uint32_t>(ceil((float)width_ / voxel_chunk_size_));
+    chunks_height_ = static_cast<uint32_t>(ceil((float)height_ / voxel_chunk_size_));
+    chunks_depth_ = static_cast<uint32_t>(ceil((float)depth_ / voxel_chunk_size_));
+
+    voxel_chunks_.resize(chunks_width_ * chunks_height_ * chunks_depth_);
+
+    std::stringstream string_pointer;
+    string_pointer << this;
+    voxels_directory_ = "disk_voxels_" + string_pointer.str();
+    std::filesystem::create_directory(voxels_directory_);
+}
+
+Voxelizer::Voxelizer(std::vector<std::byte> &&voxels, uint32_t width, uint32_t height, uint32_t depth) : 
+    width_{width},
+    height_{height},
+    depth_{depth}
+{
+    voxel_chunks_.push_back(voxels);
```

Comment:
1. this should be emplace_back instead of push_back
2. Does this properly initialize voxel_chunks_? The previous constructor populates voxel_chunks_ with an (empty) vector per possible linearized chunk index. This just populates the vector with one item - what happens if width, height, and depth are larger than voxel_chunk_size_? I'm not sure this constructor is needed in general.

Reply:
It's not really necessary, it's just there so that I could insert the voxelizer class in convert_model.cpp without voxelizing a model twice - it'll be removed once we resolve how to best use the voxelizer class there (since it seems like some of the formats use raw_voxelize_obj directly).

Reply:
I see - we should re-write convert_model to not a priori voxelize the triangle model into a raw voxel grid when using a custom format - that should only be done for the "normal" formats. For the hybrid formats, we should be using the `Voxelizer(std::string_view filepath, float voxel_size);` constructor.